File: 1905/solution.py
class Solution:
    def countSubIslands(self, grid1: list[list[int]], grid2: list[list[int]]) -> int:
        res = 0
        visited = set()
        ROWS, COLS = len(grid1), len(grid1[0])
        directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        for row in range(ROWS):
            for col in range(COLS):
                if (row, col) in visited or grid2[row][col] == 0:
                    continue
                islands = []
                islands.append([row, col])
                sub = True
                while islands:
                    r, c = islands.pop()
                    if (r, c) not in visited:
                        visited.add((r, c))
                        if grid1[r][c] == 0:
                            sub = False
                        for x, y in directions:
                            nr, nc = r + x, c + y
                            if (
                                0 <= nr < ROWS
                                and 0 <= nc < COLS
                                and (nr, nc) not in visited
                                and grid2[nr][nc] == 1
                            ):
                                islands.append([nr, nc])
                if sub:
                    res += 1
        return res

File: 1905/solution.ts
function countSubIslands(grid1: number[][], grid2: number[][]): number {
  const ROWS = grid1.length;
  const COLS = grid1[0].length;
  const dirs = [
    [1, 0],
    [0, 1],
    [-1, 0],
    [0, -1],
  ];
  const visited = new Array(ROWS)
    .fill(0)
    .map((el) => new Array(COLS).fill(false));
  let islands = 0;
  const dfs = (row: number, col: number): boolean => {
    if (
      row < 0 ||
      row == ROWS ||
      col < 0 ||
      col == COLS ||
      grid2[row][col] === 0 ||
      visited[row][col]
    ) {
      return true;
    }
    visited[row][col] = true;
    let check = true;
    if (grid1[row][col] === 0) {
      check = false;
    }

    for (const [x, y] of dirs) {
      if (!dfs(row + x, col + y)) {
        check = false;
      }
    }
    return check;
  };

  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLS; col++) {
      if (!visited[row][col] && grid2[row][col] === 1 && dfs(row, col)) {
        islands++;
      }
    }
  }
  return islands;
}
